pkg/cloudprovider/provider/scaleway/provider.go

```diff
@@ -0,0 +1,437 @@
+	cloudInstance "github.com/kubermatic/machine-controller/pkg/cloudprovider/instance"
+	scalewaytypes "github.com/kubermatic/machine-controller/pkg/cloudprovider/provider/scaleway/types"
+	cloudprovidertypes "github.com/kubermatic/machine-controller/pkg/cloudprovider/types"
+	"github.com/kubermatic/machine-controller/pkg/providerconfig"
+	providerconfigtypes "github.com/kubermatic/machine-controller/pkg/providerconfig/types"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
+)
+
+type provider struct {
+	configVarResolver *providerconfig.ConfigVarResolver
+}
+
+// New returns a Scaleway provider
+func New(configVarResolver *providerconfig.ConfigVarResolver) cloudprovidertypes.Provider {
+	return &provider{configVarResolver: configVarResolver}
+}
+
+type Config struct {
+	AccessKey      string
+	SecretKey      string
+	ProjectID      string
+	Zone           string
+	CommercialType string
+	IPv6           bool
+	Tags           []string
+}
+
+func (c *Config) getInstanceAPI() (*instance.API, error) {
+	client, err := scw.NewClient(
+		scw.WithAuth(c.AccessKey, c.SecretKey),
+		scw.WithDefaultZone(scw.Zone(c.Zone)),
+		scw.WithDefaultProjectID(c.ProjectID),
+		scw.WithUserAgent("kubermatic/machine-controller"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return instance.NewAPI(client), nil
+}
+
+func getImageNameForOS(os providerconfigtypes.OperatingSystem) (string, error) {
+	switch os {
+	case providerconfigtypes.OperatingSystemUbuntu:
+		// ubuntu_focal doesn't work (see https://bugs.launchpad.net/ubuntu/+source/linux-kvm/+bug/1880522)
+		// modprobe ip_vs will fail
+		return "ubuntu_bionic", nil
+	case providerconfigtypes.OperatingSystemCentOS:
+		return "centos_7.6", nil
+	}
+	return "", providerconfigtypes.ErrOSNotSupported
+}
+
+func (p *provider) getConfig(s v1alpha1.ProviderSpec) (*Config, *providerconfigtypes.Config, error) {
+	if s.Value == nil {
+		return nil, nil, fmt.Errorf("machine.spec.providerconfig.value is nil")
+	}
+	pconfig := providerconfigtypes.Config{}
+	err := json.Unmarshal(s.Value.Raw, &pconfig)
+	if err != nil {
+		return nil, nil, err
+	}
+	rawConfig := scalewaytypes.RawConfig{}
+	err = json.Unmarshal(pconfig.CloudProviderSpec.Raw, &rawConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c := Config{}
+	c.AccessKey, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.AccessKey, scw.ScwAccessKeyEnv)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"access_key\" field, error = %v", err)
+	}
+	c.SecretKey, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.SecretKey, scw.ScwSecretKeyEnv)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"secret_key\" field, error = %v", err)
+	}
+	c.ProjectID, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.ProjectID, scw.ScwDefaultProjectIDEnv)
```

Comment:
So no support for env variables ? 🤔

Reply:
There is a support for it, but in the context of project-id which can be added explicitly I think it's not needed, however, for the accessKey and secretKy then yeah env var are created out of the secret and used.

Reply:
Hum I was more thinking about starting the machine-controller with these env, thus not needing to specify them here, but maybe my approach is wrong 😄

Reply:
Well, this field doesn't have some sensitive information(for example custom images in Azure aren't credentials but they have some information about the subscription thus it shouldn't be added explicitly), so it can be just added explicitly in the template. Reading from the env vars means it needs to be added to the secret, however, project-id(as long as it doesn't have sensitive information) shouldn't be treated as secrets.